XXX-staged-require-with-lifecycle-hooks.md

```diff
@@ -0,0 +1,330 @@
+  resolve({ inputPath }) {
+    if (!/^config\:/.test(inputPath)) return
+    return `${this.base}/${path.slice(7)}.json`
+  }
+
+  // Return a string to set `contents` to pass along to subsequent stages
+  // Return null to remove loader from remainder of this request pipeline
+  // Throw to indicate module loading failure (other than not found)
+  load({ resolvedPath }) {
+    try {
+      return fs.readFileSync(resolvedPath).toString()
+    } catch (err) {
+      if (err.code === 'ENOENT') return
+      throw new Error(`module load error: ${err.message}`)
+    }
+  }
+
+  // Return transformed contents
+  // Throw to indicate module transform failure
+  transform({ contents, resolvedPath }) {
+    if (!contents.contains('testing')) {
+      throw new Error('module transform error: "testing" not found')
+    }
+
+    return contents.replace(/testing/, 'test')
+  }
+
+  // Return the final compiled module object
+  // Throw to indicate parse/compile failure
+  compile({ contents, resolvedPath }) {
+    try {
+      const module = new Module()
+      module.filename = resolvedPath
+      module.exports = JSON.parse(contents)
+      return module
+    } catch (err) {
+      throw new Error(`module compile error: ${err.message}`)
+    }
+  }
+}
+
+// Add the custom loader to the loaders array
+require.loaders.push(new JSONConfigLoader({
+  base: path.join(process.cwd(), 'configs')
+}))
+
+const myConfig = require('config:some-custom-config')
+```
+
+This example is implemented as a class, but it could also just be a simple
+object.
+
+Each stage handler would receive a `ModuleRequest` object, with each handler
+corresponding to an assignment to a particular property of that request. The
+`resolve` method sets `resolvedPath`, the `load` method sets `contents`, the
+`transform` method alters `contents`, the `compile` method sets `module`.
+
+Returning `null` from any of `resolve`, `load`, `transform` or `compile` will
+result in that loader being removed from the remainder of the pipeline for that
+load request. However, any of `load`, `transform` and `compile` can be left
+unimplemented to skip just that step. Returning a value from `load` will skip
+the remaining loaders in the `load` stage and proceed to the `transform` stage.
+Similar to `load`, the first `compile` to return a non-null value will skip the
+remainder of the compile step and that value will become the final module.
+
+# Notes
+
+- The `ModuleRequest` object could possibly also just be a `Module` instance,
+but the current resolution stage occurs before attempting to build a module
+instance, so it's possible backwards-incompatible changes may be required to
+do that, thus the safer approach of an entirely new object.
+- For the object-based style, it might also be a good idea to, rather than have
+the `transform` method, have `preCompile` and `postCompile` transformers.
+
+[1]: https://github.com/nodejs/node/blob/master/lib/_tls_wrap.js#L829
```

Comment:
You probably wanna future proof this link by tying it to a specific git ref (it's already boken) - tip: press `y` while on the github page and it will update the URL for you with a ref